### PR TITLE
[hermes] Introduce severity based colored console reporter

### PR DESCRIPTION
**Pull request by SlavikMIPT**
This commit introduces severity based colored console reporter.

ONE-DCO-1.0-Signed-off-by: Vyacheslav Bazhenov <slavikmipt@gmail.com>

### CODE REVIEW

**Comment by SlavikMIPT** [comment]
image

File: compiler/hermes/include/hermes/core/Message.h [hunk 1]
```diff
@@ -21,6 +21,7 @@
 #include <sstream>
 #include <string>
 #include <vector>
+#include "Severity.h"
```
Comment (binarman) [inline compiler/hermes/include/hermes/core/Message.h:24]: I think this is considered as a separate component ("hermes" vs "hermes-std" ), could you move this in separate PR?

**Comment by binarman** [comment]
@SlavikMIPT 
This PR is based on broken master, rebase should fix tests

File: compiler/hermes-std/src/ConsoleReporter.cpp [hunk 2]
```diff
@@ -20,13 +20,48 @@
 
 namespace hermes
 {
+static constexpr const char *kTermColorRedTextCode = "\033[0;31m";
```
Comment (seanshpark) [inline compiler/hermes-std/src/ConsoleReporter.cpp:23]: please put an empty line before this line

File: compiler/hermes-std/src/ConsoleReporter.test.cpp [hunk 3]
```diff
@@ -43,3 +43,83 @@ TEST(ConsoleReporterTest, notify)
 
   ASSERT_NO_THROW(r.notify(&m));
 }
+
+TEST(ConsoleReporterTest, notify_fatal)
```
Comment (seanshpark) [inline compiler/hermes-std/src/ConsoleReporter.test.cpp:47]: Please add same amout of negative tests, name that end with `_NEG`.
Refer https://github.com/Samsung/ONE/pull/7937#issuecomment-962942396 why.

File: compiler/hermes/include/hermes/core/Message.h (hunk 1 above)
Comment (seanshpark) [reply compiler/hermes/include/hermes/core/Message.h:24]: Please put this line above `#include <memory>` with an empty line separation.

**Review by seanshpark — COMMENTED** [review]
.

File: compiler/hermes-std/src/ConsoleReporter.cpp [hunk 4]
```diff
@@ -20,13 +20,48 @@
...
+static constexpr const char *kTermInverseOffTextCode = "\033[27m";
+
+static constexpr const char *kTermColorResetAllCode = "\033[0m";
 
 void ConsoleReporter::notify(const hermes::Message *m)
 {
+  switch (m->get_severity())
+  {
+    case FATAL:
+      std::cout << kTermColorRedTextCode << kTermBoldTextCode << kTermUnderlineTextCode;
```
Comment (seanshpark) [inline compiler/hermes-std/src/ConsoleReporter.cpp:45]: If these eacape character are to be show, these may affect tools that may use log captures.
Pleae provide an option to turn this on, where default is off.
- a method, that can be set from the caller of hermes
- environment variable, such as `ONE_HERMES_COLOR=1`
- and maybe more...

File: compiler/hermes-std/src/ConsoleReporter.cpp (hunk 2 above)
Comment (seanshpark) [inline compiler/hermes-std/src/ConsoleReporter.cpp:23]: Please put an empty line before this line

Comment (seanshpark) [inline compiler/hermes-std/src/ConsoleReporter.cpp:23]: Please put an empty line before this line

Comment (seanshpark) [inline compiler/hermes-std/src/ConsoleReporter.cpp:23]: Please put an empty line before this line

Comment (seanshpark) [inline compiler/hermes-std/src/ConsoleReporter.cpp:23]: Please put an empty line before this line

File: compiler/hermes-std/src/ConsoleReporter.cpp (hunk 4 above)
Comment (seanshpark) [inline compiler/hermes-std/src/ConsoleReporter.cpp:45]: If these eacape character are to be shown, these may affect tools that may use log captures.
Pleae provide an option to turn this on, where default is off.
- a method, that can be set from the caller of hermes
- environment variable, such as ONE_HERMES_COLOR=1
- and maybe more...

File: compiler/hermes/include/hermes/core/Message.h [hunk 5]
```diff
@@ -57,11 +57,17 @@ class Message final
   Message() = default;
 
 public:
-  void text(std::unique_ptr<MessageText> &&text) { _text = std::move(text); }
+  void text(std::unique_ptr<MessageText> &&text, SeverityCategory severity = VERBOSE)
```
Comment (seanshpark) [inline compiler/hermes/include/hermes/core/Message.h:60]: Please do not use = VERBOSE. We don't use implicit parameter settings.

File: compiler/hermes/include/hermes/core/Message.h (hunk 1 above)
Comment (SlavikMIPT) [reply compiler/hermes/include/hermes/core/Message.h:24]: > I think this is considered as a separate component ("hermes" vs "hermes-std" ), could you move this in separate PR?

https://github.com/Samsung/ONE/pull/8002

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

File: compiler/hermes-std/src/ConsoleReporter.cpp [hunk 6]
```diff
@@ -17,16 +17,68 @@
...
+
 void ConsoleReporter::notify(const hermes::Message *m)
 {
+  const char *env_color_p = std::getenv("ONE_HERMES_COLOR");
+  if (env_color_p)
+  {
+    auto env_color_str = std::string(env_color_p);
+    if ((env_color_str == "1") or (env_color_str == "ON"))
+      _is_colored = true;
+  }
```
Comment (seanshpark) [inline compiler/hermes-std/src/ConsoleReporter.cpp:51]: (optional) (A) this routine is simple and I think it won't take that much time but will run in everly lines of message.
you can (1) add a TODO like "optimize this for speed" or (2) revise this code to check if not checked before
and (B) when turned on, there is no check to turn it off. I'm not sure when once turned one, we need to check for off but for safetly in usage case, we may need it. for this case, we may have to skip (A) case.

Anyway we can go as-is or just add some TODOs...
this library maybe used in our TVN backend and if something goes wrong, there will be some reports.

**Review by binarman — APPROVED** [review]
LGTM

File: compiler/hermes-std/src/ConsoleReporter.cpp (hunk 6 above)
Comment (binarman) [reply compiler/hermes-std/src/ConsoleReporter.cpp:51]: I've tested if this code and found that million calls takes ~20 milliseconds on my hardware.
This should not be a big problem, at least for now =)

**Review by jj-crypto — APPROVED** [review]
